src/transformers/models/mm_grounding_dino/modular_mm_grounding_dino.py

```diff
@@ -0,0 +1,258 @@
+        text_token_mask: torch.BoolTensor,
+    ) -> torch.FloatTensor:
+        res = vision_hidden_state @ text_hidden_state.transpose(-1, -2)
+        res = res / math.sqrt(vision_hidden_state.shape[-1])
+        res = res + self.bias
+        res.masked_fill_(~text_token_mask[:, None, :], float("-inf"))
+
+        # padding to max_text_len
+        new_res = torch.full((*res.shape[:-1], self.max_text_len), float("-inf"), device=res.device)
+        new_res[..., : res.shape[-1]] = res
+
+        return new_res
+
+
+class MMGroundingDinoPreTrainedModel(GroundingDinoPreTrainedModel):
+    def _init_weights(self, module):
+        super()._init_weights(module)
+        if isinstance(module, MMGroundingDinoContrastiveEmbedding):
+            nn.init.constant_(module.bias, -math.log((1 - 0.01) / 0.01))
+
+
+# TODO: this one is useless, but without it class order in modeling gets messed up
+class MMGroundingDinoDecoder(GroundingDinoDecoder):
+    pass
+
+
+class MMGroundingDinoModel(GroundingDinoModel):
+    def __init__(self, config: MMGroundingDinoConfig):
+        super().__init__(config)
+        self.encoder_output_class_embed = MMGroundingDinoContrastiveEmbedding(config)
+
+
+class MMGroundingDinoMLPPredictionHead(GroundingDinoMLPPredictionHead):
+    pass
+
+
+class MMGroundingDinoForObjectDetection(GroundingDinoForObjectDetection, MMGroundingDinoPreTrainedModel):
+    _tied_weights_keys = [
+        r"bbox_embed\.[1-9]\d*",
+        r"model\.decoder\.bbox_embed\.[0-9]\d*",
+        r"class_embed\.[1-9]\d*",
+        r"model\.decoder\.class_embed\.[0-9]\d*",
+    ]
+
+    def __init__(self, config: MMGroundingDinoConfig):
+        MMGroundingDinoPreTrainedModel.__init__(config)
+
+        self.model = MMGroundingDinoModel(config)
+
+        if config.decoder_cls_embed_share:
+            _class_embed = MMGroundingDinoContrastiveEmbedding(config)
+            self.class_embed = nn.ModuleList([_class_embed for _ in range(config.decoder_layers)])
+        else:
+            module_list = []
+            for _ in range(config.decoder_layers):
+                _class_embed = MMGroundingDinoContrastiveEmbedding(config)
+                module_list.append(_class_embed)
+            self.class_embed = nn.ModuleList(module_list)
+
+        if config.decoder_bbox_embed_share:
+            _bbox_embed = MMGroundingDinoMLPPredictionHead(
+                input_dim=config.d_model, hidden_dim=config.d_model, output_dim=4, num_layers=3
+            )
+            self.bbox_embed = nn.ModuleList([_bbox_embed for _ in range(config.decoder_layers)])
+        else:
+            module_list = []
+            for _ in range(config.decoder_layers):
+                _bbox_embed = MMGroundingDinoMLPPredictionHead(
+                    input_dim=config.d_model, hidden_dim=config.d_model, output_dim=4, num_layers=3
+                )
+                module_list.append(_bbox_embed)
+            self.bbox_embed = nn.ModuleList(module_list)
```

Comment:
in general we should not have so many code paths! Each case would lead to a different Embedding.... Are there a model for each case?

Reply:
ahh, I see! Yes, let's copy and modify the whole config class then